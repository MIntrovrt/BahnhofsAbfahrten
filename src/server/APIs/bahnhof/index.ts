import { getLageplan } from 'server/Bahnhof/Lageplan';
import KoaRouter from 'koa-router';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter().get('/lageplan/:stationName', async ctx => {
    const { stationName }: { stationName: string } = ctx.params;

    ctx.body = {
      lageplan: await getLageplan(stationName),
    };
  });

router
  .prefix('/bahnhof')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes());

export default router;
export const versions = ['current', 'v1'];
