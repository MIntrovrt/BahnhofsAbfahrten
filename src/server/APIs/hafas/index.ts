import auslastungHafas from 'server/Auslastung/Hafas';
import geoStation from 'server/HAFAS/LocGeoPos';
import journeyDetails from 'server/HAFAS/JourneyDetails';
import KoaRouter from 'koa-router';
import LocMatch from 'server/HAFAS/LocMatch';
import makeRequest from 'server/HAFAS/Request';
import routing from 'server/HAFAS/TripSearch';
import stationBoard from 'server/HAFAS/StationBoard';
import trainSearch from 'server/HAFAS/TrainSearch';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter()
    .get('/journeyDetails/:jid', async ctx => {
      const { jid }: { jid: string } = ctx.params;

      ctx.body = await journeyDetails(jid);
    })
    .get('/details/:trainName/:date', async ctx => {
      const { date, trainName } = ctx.params;

      const trains = await trainSearch(trainName, Number.parseInt(date, 10));

      if (trains.length) {
        ctx.body = await journeyDetails(trains[0].jid);
      } else {
        ctx.status = 404;
      }
    })
    .get('/auslastung/:start/:dest/:trainNumber/:time', async ctx => {
      const { start, dest, time, trainNumber } = ctx.params;

      ctx.body = await auslastungHafas(
        start,
        dest,
        trainNumber,
        Number.parseInt(time, 10)
      );
    })
    .get('/ArrStationBoard', async ctx => {
      const { date, station } = ctx.query;

      ctx.body = await stationBoard({
        type: 'ARR',
        station,
        date: Number.parseInt(date, 10) || undefined,
      });
    })
    .get('/DepStationBoard', async ctx => {
      const { date, station } = ctx.query;

      ctx.body = await stationBoard({
        type: 'DEP',
        station,
        date: Number.parseInt(date, 10) || undefined,
      });
    })
    .get('/trainSearch/:trainName/:date', async ctx => {
      const { date, trainName } = ctx.params;

      ctx.body = await trainSearch(trainName, Number.parseInt(date, 10));
    })
    .get('/geoStation', async ctx => {
      const { x, y, lat, lng, maxDist = 1000 } = ctx.query;

      const realY = lat ? Number.parseFloat(lat) * 1000000 : y;
      const realX = lng ? Number.parseFloat(lng) * 1000000 : x;

      ctx.body = await geoStation(realX, realY, maxDist);
    })
    .get('/station/:searchTerm', async ctx => {
      const { searchTerm } = ctx.params;

      ctx.body = await LocMatch(searchTerm);
    })
    .post('/rawHafas', async ctx => {
      ctx.body = await makeRequest(ctx.request.body);
    })
    .post('/route', async ctx => {
      ctx.body = await routing({
        ...ctx.request.body,
        time: Number.parseInt(ctx.request.body.time, 10),
      });
    });

router
  .prefix('/hafas')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes());

export default router;
export const versions = ['v1'];
