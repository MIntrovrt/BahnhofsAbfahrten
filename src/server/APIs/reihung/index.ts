import {
  wagenReihung,
  wagenReihungMonitoring,
  wagenReihungStation,
} from 'server/Reihung';
import KoaRouter from 'koa-router';

const router = new KoaRouter();
const getCurrent = () =>
  new KoaRouter()
    .get('/wagenstation/:train/:station', async ctx => {
      const { train, station }: { train: string; station: string } = ctx.params;

      ctx.body = await wagenReihungStation([train], station);
    })
    .get('/wagen/:trainNumber/:date', async ctx => {
      const {
        date,
        trainNumber,
      }: { date: string; trainNumber: string } = ctx.params;

      ctx.body = await wagenReihung(trainNumber, Number.parseInt(date, 10));
    });

router
  .prefix('/reihung')
  .use('/current', getCurrent().routes())
  .use('/v1', getCurrent().routes())
  .get('/monitoring/wagen', async ctx => {
    ctx.body = await wagenReihungMonitoring();
    if (!ctx.body) {
      ctx.status = 404;
    }
  });

export default router;
export const versions = ['v1'];
