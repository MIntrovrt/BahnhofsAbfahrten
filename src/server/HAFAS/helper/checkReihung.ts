import { TrnCmpSX } from 'types/HAFAS';

const allowdTypes = ['ICE', 'IC', 'TGV', 'EC', 'ECE', 'RJ', 'D'];

export default (trnCmpSX?: TrnCmpSX, trainType?: string) => {
  if (trainType && allowdTypes.includes(trainType)) return true;

  if (trnCmpSX && trnCmpSX.tcM) return true;
};
