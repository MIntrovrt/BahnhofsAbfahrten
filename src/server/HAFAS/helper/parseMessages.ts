import { MsgL, ParsedCommon } from 'types/HAFAS';

export default (msgL: MsgL[], common: ParsedCommon) => {
  const messages = msgL.map(msg => common.remL[msg.remX]);

  return messages;
};
