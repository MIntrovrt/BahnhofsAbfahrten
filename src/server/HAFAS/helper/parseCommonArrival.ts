import { CommonArrival, ParsedCommon } from 'types/HAFAS';
import { differenceInMinutes } from 'date-fns';
import { ParsedCommonArrival } from 'types/common';
import parseMessages from './parseMessages';
import parseTime from './parseTime';

export default (
  a: CommonArrival,
  date: number,
  common: ParsedCommon,
  trainType?: string
): ParsedCommonArrival => {
  const scheduledTime = parseTime(date, a.aTimeS);
  let time = scheduledTime;
  let delay;

  if (a.aTimeR) {
    time = parseTime(date, a.aTimeR);
    delay = time && scheduledTime && differenceInMinutes(time, scheduledTime);
  }

  return {
    scheduledPlatform: a.aPlatfR && a.aPlatfS,
    platform: a.aPlatfR || a.aPlatfS,
    scheduledTime,
    time,
    delay,
    reihung:
      Boolean(a.aTrnCmpSX && a.aTrnCmpSX.tcM) ||
      trainType === 'IC' ||
      trainType === 'EC' ||
      undefined,
    messages: a.msgL ? parseMessages(a.msgL, common) : undefined,
  };
};
