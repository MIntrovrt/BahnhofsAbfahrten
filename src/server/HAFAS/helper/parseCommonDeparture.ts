import { CommonDeparture, ParsedCommon } from 'types/HAFAS';
import { differenceInMinutes } from 'date-fns';
import { ParsedCommonDeparture } from 'types/common';
import parseMessages from './parseMessages';
import parseTime from './parseTime';

export default (
  d: CommonDeparture,
  date: number,
  common: ParsedCommon,
  trainType?: string
): ParsedCommonDeparture => {
  const scheduledTime = parseTime(date, d.dTimeS);
  let time = scheduledTime;
  let delay;

  if (d.dTimeR) {
    time = parseTime(date, d.dTimeR);
    delay = time && scheduledTime && differenceInMinutes(time, scheduledTime);
  }

  return {
    scheduledPlatform: d.dPlatfR && d.dPlatfS,
    platform: d.dPlatfR || d.dPlatfS,
    scheduledTime,
    time,
    delay,
    reihung:
      Boolean(d.dTrnCmpSX && d.dTrnCmpSX.tcM) ||
      trainType === 'IC' ||
      trainType === 'EC' ||
      undefined,
    messages: d.msgL ? parseMessages(d.msgL, common) : undefined,
  };
};
