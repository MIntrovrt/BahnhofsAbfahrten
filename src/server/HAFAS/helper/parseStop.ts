import { CommonStop, ParsedCommon } from 'types/HAFAS';
import { Route$Stop } from 'types/routing';
import parseAuslastung from './parseAuslastung';
import parseCommonArrival from './parseCommonArrival';
import parseCommonDeparture from './parseCommonDeparture';

export default (
  stop: CommonStop,
  common: ParsedCommon,
  date: number,
  trainType?: string
): Route$Stop => ({
  station: common.locL[stop.locX],
  arrival: stop.aTimeS
    ? parseCommonArrival(stop, date, common, trainType)
    : undefined,
  departure: stop.dTimeS
    ? parseCommonDeparture(stop, date, common, trainType)
    : undefined,
  auslastung: parseAuslastung(stop.dTrnCmpSX, common.tcocL),
});
