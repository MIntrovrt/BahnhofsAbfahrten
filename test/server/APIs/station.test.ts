import { checkApi, mockWithFile } from './helper';
import { versions } from 'server/APIs/station';
jest.mock('node-cache');

describe('Station API', () => {
  versions.forEach(v => {
    describe(v, () => {
      it('/search favendo', async () => {
        mockWithFile(
          'https://si.favendo.de',
          '/station-info/rest/api/search?searchTerm=Hamburg+Hbf',
          'station/favendo'
        );
        await checkApi(`/api/station/${v}/search/Hamburg Hbf?type=1`);
      });
      it('/search openData', async () => {
        mockWithFile(
          'https://api.deutschebahn.com',
          '/stada/v2/stations?searchstring=*Hamburg*Hbf*',
          'station/opendata'
        );
        await checkApi(`/api/station/${v}/search/Hamburg Hbf?type=3`);
      });
      it('/search openData offline', async () => {
        await checkApi(`/api/station/${v}/search/Hamburg Hbf?type=4`);
      });
      it('/iris', async () => {
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8002549',
          'station/iris/1'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8071065',
          'station/iris/2'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8071066',
          'station/iris/3'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8076116',
          'station/iris/4'
        );
        mockWithFile(
          'https://iris.noncd.db.de',
          '/iris-tts/timetable/station/8098549',
          'station/iris/5'
        );
        await checkApi(`/api/station/${v}/iris/8002549`);
      });
    });
  });
});
