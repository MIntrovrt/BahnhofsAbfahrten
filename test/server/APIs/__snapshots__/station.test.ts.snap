// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Station API v1 /iris 1`] = `
Object {
  "relatedStations": Array [
    Object {
      "creationts": "19-06-09 02:14:17.278",
      "db": "true",
      "ds100": "AH  K",
      "eva": "8071065",
      "meta": Array [
        "8002549",
        "8071066",
        "8076116",
      ],
      "name": "Hamburg Hbf (Kirchenallee)",
    },
    Object {
      "creationts": "19-06-09 02:14:17.159",
      "db": "true",
      "ds100": "AHS",
      "eva": "8098549",
      "meta": Array [
        "8002549",
        "8076116",
      ],
      "name": "Hamburg Hbf (S-Bahn)",
    },
    Object {
      "creationts": "19-06-09 02:14:17.278",
      "db": "true",
      "ds100": "AH  S",
      "eva": "8071066",
      "meta": Array [
        "8002549",
        "8071065",
        "8076116",
      ],
      "name": "Hamburg Hbf (Spitalerstra√üe)",
    },
    Object {
      "creationts": "19-06-09 02:14:17.508",
      "db": "true",
      "ds100": "AH  Z",
      "eva": "8076116",
      "meta": Array [
        "8002549",
        "8071065",
        "8071066",
        "8098549",
      ],
      "name": "Hamburg Hbf ZOB",
    },
  ],
  "station": Object {
    "creationts": "19-06-09 02:14:16.951",
    "db": "true",
    "ds100": "AH",
    "eva": "8002549",
    "meta": Array [
      "8071065",
      "8071066",
      "8076116",
      "8098549",
    ],
    "name": "Hamburg Hbf",
    "p": "13b|11a|11b|5a/b|7a/b|7a|7b|5a|5b|S|13a/b|11a/b|14a|14b|11|12a|12|12b|13|14|8a/b|6a/b|8a|8b|6a|6b|5|6|7|8|12a/b|14a/b|13a",
  },
}
`;

exports[`Station API v1 /search favendo 1`] = `
Array [
  Object {
    "favendoId": 2514,
    "id": "8002549",
    "title": "Hamburg Hbf",
  },
]
`;

exports[`Station API v1 /search openData 1`] = `
Array [
  Object {
    "DS100": "AH",
    "favendoId": 2514,
    "id": "8002549",
    "title": "Hamburg Hbf",
  },
]
`;

exports[`Station API v1 /search openData offline 1`] = `
Array [
  Object {
    "DS100": "AH",
    "favendoId": 2514,
    "id": "8002549",
    "title": "Hamburg Hbf",
  },
  Object {
    "DS100": "SHO",
    "favendoId": 2892,
    "id": "8000176",
    "title": "Homburg (Saar) Hbf",
  },
  Object {
    "DS100": "UNM",
    "favendoId": 4309,
    "id": "8010240",
    "title": "Naumburg (Saale) Hbf",
  },
]
`;
